src/Compiler/Optimize/Optimizer.fs

```diff
@@ -2898,10 +2903,11 @@ and OptimizeLinearExpr cenv env expr contf =
 
       let (bindR, bindingInfo), env = OptimizeBinding cenv false env bind 
 
-      OptimizeLinearExpr cenv env body (contf << (fun (bodyR, bodyInfo) ->  
+      OptimizeLinearExpr cenv env body (contf << (fun (bodyR, bodyInfo) -> 
         // PERF: This call to ValueIsUsedOrHasEffect/freeInExpr amounts to 9% of all optimization time.
         // Is it quadratic or quasi-quadratic?
-        if ValueIsUsedOrHasEffect cenv (fun () -> (freeInExpr (CollectLocalsWithStackGuard()) bodyR).FreeLocals) (bindR, bindingInfo) then
+        let collect expr = (freeInExpr (CollectLocalsWithStackGuard()) expr).FreeLocals
+        if ValueIsUsedOrHasEffect cenv (fun () -> (getFreeLocalsCache cenv).GetOrAdd(bodyR, collect)) (bindR, bindingInfo) then
```

Comment:
This gives a modest 20% hit ratio in tests. May or may not help somewhat IRL but needs benchmarking.

Reply:
Looking at Optimization times when building FCS I don't see any speedup.